research/flamby/fed_isic2019/ditto/client.py

```diff
@@ -0,0 +1,150 @@
+import argparse
+import os
+from logging import INFO
+from pathlib import Path
+from typing import Dict, Optional, Sequence, Tuple
+
+import flwr as fl
+import torch
+import torch.nn as nn
+from flamby.datasets.fed_isic2019 import BATCH_SIZE, LR, NUM_CLIENTS, Baseline, BaselineLoss
+from flwr.common.logger import log
+from flwr.common.typing import Config
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+
+from fl4health.checkpointing.checkpointer import BestMetricTorchCheckpointer, TorchCheckpointer
+from fl4health.clients.ditto_client import DittoClient
+from fl4health.utils.losses import LossMeterType
+from fl4health.utils.metrics import BalancedAccuracy, Metric
+from fl4health.utils.random import set_all_random_seeds
+from research.flamby.flamby_data_utils import construct_fedisic_train_val_datasets
+
+
+class FedIsic2019DittoClient(DittoClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        client_number: int,
+        learning_rate: float,
+        lam: float,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+            lam=lam,
+        )
+        self.client_number = client_number
+        self.learning_rate: float = learning_rate
+
+        assert 0 <= client_number < NUM_CLIENTS
+        log(INFO, f"Client Name: {self.client_name}, Client Number: {self.client_number}")
+
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        train_dataset, validation_dataset = construct_fedisic_train_val_datasets(
+            self.client_number, str(self.data_path)
+        )
+        train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
+        val_loader = DataLoader(validation_dataset, batch_size=BATCH_SIZE, shuffle=False)
+        return train_loader, val_loader
+
+    def get_model(self, config: Config) -> nn.Module:
+        model: nn.Module = Baseline().to(self.device)
```

Comment:
I might be totally wrong, but are we training the same number of parameters in Ditto as in other methods (similar to the case in APFL or FENDA)? Given that in Ditto, we are training both a global and local model in each client, should the model size be approximately half of the baseline model? 
If that's true, we can probably achieve this by using the APFL model structure.

Reply:
I think this is a good discussion point and thought. I considered this when writing the FedISIC scripts and would be interested in hearing both your thoughts. The model being trained isn't "twice as large" in that it has the same number of parameters as a centralized method (similar to FedPer). So it doesn't have additional expressiveness w.r.t the number of DoFs. However, we are technically training two models of the same size (global and local, but only one is used in prediction, as opposed to APFL). So the **cost of training** is higher for Ditto but the model isn't bigger. We should definitely method at least mention this in the paper discussion either way.

Reply:
Right, I think this makes sense, since only one model is used for prediction, and the other is trained to facilitate the main one (local model) in the training process. So, I will also use the full model for GEMINI experiments if John has the same opinion.

Reply:
I think your raise a good point Fatemeh! But I agree with David that we shouldn't consider the global models parameter counts since at inference only the local model is used. Maybe SCAFFOLD is a somewhat useful analogy (although not perfect obviously), we don't consider control variates in the parameter count. That said, definitely worth mentioning the additional cost of training in the paper

Reply:
Yeah SCAFFOLD is a good example! 👍 I will also use the full model then.